lib/mysql_framework/connector.rb

```diff
@@ -3,19 +3,50 @@
 module MysqlFramework
   class Connector
     def initialize(options = {})
-      @connection_pool = ::Queue.new
-
       @options = default_options.merge(options)
 
       Mysql2::Client.default_query_options.merge!(symbolize_keys: true, cast_booleans: true)
     end
 
-    # This method is called to fetch a client from the connection pool or create a new client if no idle clients
-    # are available.
+    # This method is called to setup a pool of MySQL connections.
+    def setup
+      @connection_pool = ::Queue.new
+
+      start_pool_size.times { @connection_pool.push(Mysql2::Client.new(@options)) }
+
+      @created_connections = start_pool_size
+    end
+
+    # This method is called to close all MySQL connections in the pool and dispose of the pool itself.
+    def dispose
+      return if @connection_pool.nil?
+
+      until @connection_pool.empty?
+        conn = @connection_pool.pop(true)
+        conn.close
+      end
+
+      @connection_pool = nil
+    end
+
+    # This method is called to get the idle connection queue for this connector.
+    def connections
+      @connection_pool
+    end
+
+    # This method is called to fetch a client from the connection pool.
     def check_out
       @connection_pool.pop(true)
-    rescue StandardError
-      Mysql2::Client.new(@options)
+    rescue ThreadError
+      if @created_connections < max_pool_size
+        conn = Mysql2::Client.new(@options)
+        @created_connections += 1
+        return conn
+      end
+
+      MysqlFramework.logger.error { "[#{self.class}] - Database connection pool depleted." }
+
+      raise 'Database connection pool depleted.'
```

Comment:
Did you decide to raise an error here instead of making a blocking call with a timeout when no connection is available?

Reply:
Yes - after our conversation yesterday I attempted to implement a retry mechanism and the blocking call. 

After experimenting, I felt raising an explicit expception here that could be handled by the implemeting app was the best way to go - it at least imforms implementers that either their connection pool is too small, or they're not handling their connections properly and checking them back into the pool when done.